File: taskstatus.go
package main

import (
	"fmt"
	"log"
	"strconv"
	"sync"
	"time"

	"github.com/taskcluster/taskcluster-client-go"
	"github.com/taskcluster/taskcluster-client-go/tcqueue"
)

// Enumerate task status to aid life-cycle decision making
// Use strings for benefit of simple logging/reporting
const (
	unclaimed        TaskStatus = "Unclaimed"
	claimed          TaskStatus = "Claimed"
	reclaimed        TaskStatus = "Reclaimed"
	aborted          TaskStatus = "Aborted"
	cancelled        TaskStatus = "Cancelled"
	succeeded        TaskStatus = "Succeeded"
	failed           TaskStatus = "Failed"
	errored          TaskStatus = "Errored"
	unknown          TaskStatus = "Unknown"
	deadlineExceeded TaskStatus = "Deadline Exceeded"
)

const (
	workerShutdown      TaskUpdateReason = "worker-shutdown"
	malformedPayload    TaskUpdateReason = "malformed-payload"
	resourceUnavailable TaskUpdateReason = "resource-unavailable"
	internalError       TaskUpdateReason = "internal-error"
	superseded          TaskUpdateReason = "superseded"
	intermittentTask    TaskUpdateReason = "intermittent-task"
)

type TaskStatusChangeListener struct {
	Name     string
	Callback func(ts TaskStatus)
}

type TaskStatusManager struct {
	sync.Mutex
	task       *TaskRun
	takenUntil tcclient.Time
	status     tcqueue.TaskStatusStructure
	// callback functions to call when status changes
	statusChangeListeners map[*TaskStatusChangeListener]bool
	abortException        *CommandExecutionError
	// closed when reclaim go routine should stop reclaiming
	stopReclaiming chan<- struct{}
	// closed when reclaim loop exits
	reclaimingDone <-chan struct{}
	// true if reclaims are no longer taking place for this task
	finishedReclaiming bool
}

func (tsm *TaskStatusManager) DeregisterListener(listener *TaskStatusChangeListener) {
	delete(tsm.statusChangeListeners, listener)
}

func (tsm *TaskStatusManager) RegisterListener(listener *TaskStatusChangeListener) {
	tsm.Lock()
	defer tsm.Unlock()
	tsm.statusChangeListeners[listener] = true
}

type TaskStatusUpdateError struct {
	Message       string
	CurrentStatus TaskStatus
}

func (ue *TaskStatusUpdateError) Error() string {
	return ue.Message + " (current status: " + string(ue.CurrentStatus) + ")"
}

func (tsm *TaskStatusManager) ReportException(reason TaskUpdateReason) error {
	return tsm.updateStatus(
		errored,
		func(task *TaskRun) error {
			tsm.stopReclaims()
			ter := tcqueue.TaskExceptionRequest{Reason: string(reason)}
			tsr, err := task.Queue.ReportException(task.TaskID, strconv.FormatInt(int64(task.RunID), 10), &ter)
			if err != nil {
				log.Printf("Not able to report exception for task %v:", task.TaskID)
				log.Printf("%v", err)
				return err
			}
			tsm.status = tsr.Status
			tsm.takenUntil = tcclient.Time{}
			return nil
		},
		claimed,
		reclaimed,
		aborted,
	)
}

func (tsm *TaskStatusManager) ReportFailed() error {
	return tsm.updateStatus(
		failed,
		func(task *TaskRun) error {
			tsm.stopReclaims()
			tsr, err := task.Queue.ReportFailed(task.TaskID, strconv.FormatInt(int64(task.RunID), 10))
			if err != nil {
				log.Printf("Not able to report failed completion for task %v:", task.TaskID)
				log.Printf("%v", err)
				return err
			}
			tsm.status = tsr.Status
			tsm.takenUntil = tcclient.Time{}
			return nil
		},
		claimed,
		reclaimed,
		aborted,
	)
}

func (tsm *TaskStatusManager) ReportCompleted() error {
	return tsm.updateStatus(
		succeeded,
		func(task *TaskRun) error {
			tsm.stopReclaims()
			log.Printf("Task %v finished successfully!", task.TaskID)
			tsr, err := task.Queue.ReportCompleted(task.TaskID, strconv.FormatInt(int64(task.RunID), 10))
			if err != nil {
				log.Printf("Not able to report successful completion for task %v:", task.TaskID)
				log.Printf("%v", err)
				return err
			}
			tsm.status = tsr.Status
			tsm.takenUntil = tcclient.Time{}
			return nil
		},
		claimed,
		reclaimed,
	)
}

func (tsm *TaskStatusManager) reclaim() error {
	return tsm.updateStatus(
		reclaimed,
		func(task *TaskRun) error {
			log.Printf("Reclaiming task %v...", task.TaskID)
			tcrsp, err := task.Queue.ReclaimTask(task.TaskID, fmt.Sprintf("%d", task.RunID))

			// check if an error occurred...
			if err != nil {
				// probably task was cancelled - in any case, we should kill the running task...
				log.Printf("%v", err)
				task.kill()
				return err
			}

			task.TaskReclaimResponse = *tcrsp
			// Don't need a mutex here, since tsm.updateStatus is already mutex-protected
			task.Queue = tcqueue.New(&tcclient.Credentials{
				ClientID:    tcrsp.Credentials.ClientID,
				AccessToken: tcrsp.Credentials.AccessToken,
				Certificate: tcrsp.Credentials.Certificate,
			})
			tsm.status = tcrsp.Status
			tsm.takenUntil = tcrsp.TakenUntil
			if err != nil {
				log.Printf("SERIOUS BUG: invalid credentials in queue claim response body: %v", err)
			}
			log.Printf("Reclaimed task %v successfully.", task.TaskID)
			return nil
		},
		claimed,
		reclaimed,
	)
}

func (tsm *TaskStatusManager) AbortException() *CommandExecutionError {
	tsm.Lock()
	defer tsm.Unlock()
	return tsm.abortException
}

func (tsm *TaskStatusManager) TakenUntil() tcclient.Time {
	tsm.Lock()
	defer tsm.Unlock()
	return tsm.takenUntil
}

func (tsm *TaskStatusManager) Abort(cee *CommandExecutionError) error {
	return tsm.updateStatus(
		aborted,
		func(task *TaskRun) error {
			task.Errorf("Aborting task...")
			task.kill()
			tsm.abortException = cee
			return nil
		},
		claimed,
		reclaimed,
	)
}

func (tsm *TaskStatusManager) Cancel() error {
	return tsm.updateStatus(
		cancelled,
		func(task *TaskRun) error {
			//TODO: implement cancelling of tasks
			return nil
		},
		claimed,
		reclaimed,
		aborted,
	)
}

func (tsm *TaskStatusManager) LastKnownStatus() TaskStatus {
	tsm.Lock()
	defer tsm.Unlock()
	return tsm.task.Status
}

// Queries the queue to get the latest status. Note, it can't recognise
// internal states claimed/reclaimed/aborted but is useful for setting
// failed/cancelled/pending/completed/exception.
func (tsm *TaskStatusManager) UpdateStatus() {
	tsm.Lock()
	defer tsm.Unlock()
	tsm.queryQueueForLatestStatus()
}

func (tsm *TaskStatusManager) queryQueueForLatestStatus() {
	log.Printf("Querying queue to get latest status for task %v...", tsm.task.TaskID)
	// no scopes required for this endpoint, so can use global Queue object
	// this is also useful if tsm.task.Queue == nil (which can happen if claim
	// failed because task is claimed by another worker)
	tsr, err := queue.Status(tsm.task.TaskID)
	if err != nil {
		tsm.task.Status = unknown
		return
	}

	taskStatus := tsr.Status.Runs[tsm.task.RunID]
	switch {
	case taskStatus.ReasonResolved == "failed":
		tsm.task.Status = failed
	case taskStatus.ReasonResolved == "canceled":
		tsm.task.Status = cancelled
	case taskStatus.ReasonResolved == "deadline-exceeded":
		tsm.task.Status = deadlineExceeded
	case taskStatus.State == "pending":
		tsm.task.Status = unclaimed
	case taskStatus.State == "completed":
		tsm.task.Status = succeeded
	case taskStatus.State == "exception":
		tsm.task.Status = errored
	}
	log.Printf("Latest status: %v", tsm.task.Status)
}

func (tsm *TaskStatusManager) updateStatus(ts TaskStatus, f func(task *TaskRun) error, fromStatuses ...TaskStatus) error {
	tsm.Lock()
	defer tsm.Unlock()
	currentStatus := tsm.task.Status
	for _, allowedStatus := range fromStatuses {
		if currentStatus == allowedStatus {
			e := f(tsm.task)
			if e != nil {
				tsm.queryQueueForLatestStatus()
				return &TaskStatusUpdateError{
					Message:       e.Error(),
					CurrentStatus: tsm.task.Status,
				}
			}
			tsm.task.Status = ts
			for listener := range tsm.statusChangeListeners {
				log.Printf("Notifying listener %v of state change", listener.Name)
				listener.Callback(ts)
			}
			return nil
		}
	}
	warning := fmt.Sprintf("Not updating status of task %v run %v from %v to %v. This is because you can only update to status %v if the previous status was one of: %v", tsm.task.TaskID, tsm.task.RunID, tsm.task.Status, ts, ts, fromStatuses)
	log.Print(warning)
	return &TaskStatusUpdateError{
		Message:       warning,
		CurrentStatus: tsm.task.Status,
	}
}

func NewTaskStatusManager(task *TaskRun) *TaskStatusManager {

	stopReclaiming := make(chan struct{})
	reclaimingDone := make(chan struct{})

	tsm := &TaskStatusManager{
		task:                  task,
		takenUntil:            task.TaskClaimResponse.TakenUntil,
		status:                task.TaskClaimResponse.Status,
		statusChangeListeners: map[*TaskStatusChangeListener]bool{},
		stopReclaiming:        stopReclaiming,
		reclaimingDone:        reclaimingDone,
	}

	// Reclaiming Tasks
	// ----------------
	// When the worker has claimed a task, it's said to have a claim to a given
	// `taskId`/`runId`. This claim has an expiration, see the `takenUntil`
	// property in the _task status structure_ returned from `tcqueue.ClaimTask`
	// and `tcqueue.ReclaimTask`. A worker must call `tcqueue.ReclaimTask` before
	// the claim denoted in `takenUntil` expires. It's recommended that this
	// attempted a few minutes prior to expiration, to allow for clock drift.

	go func() {
		defer close(reclaimingDone)
		for {
			var waitTimeUntilReclaim time.Duration
			if reclaimEvery5Seconds {
				// Reclaim in 5 seconds...
				waitTimeUntilReclaim = time.Second * 5
			} else {
				// Reclaim 3 mins before current claim expires...
				takenUntil := time.Time(tsm.TakenUntil())
				reclaimTime := takenUntil.Add(time.Minute * -3)
				// Round(0) forces wall time calculation instead of monotonic time in case machine slept etc
				waitTimeUntilReclaim = reclaimTime.Round(0).Sub(time.Now())
				log.Printf("Reclaiming task %v at %v", task.TaskID, reclaimTime)
				log.Printf("Current task claim expires at %v", takenUntil)
				// sanity check - only set an alarm, if wait time >= 30s, so we can't hammer queue in production
				if waitTimeUntilReclaim.Seconds() < 30 {
					log.Printf("WARNING: This is less than 30 seconds away. NOT setting a reclaim timer for task %v to avoid hammering queue if this is a bug.", task.TaskID)
					return
				}
			}
			log.Printf("Reclaiming task %v in %v", task.TaskID, waitTimeUntilReclaim)
			select {
			case <-stopReclaiming:
				return
			case <-time.After(waitTimeUntilReclaim):
				log.Printf("About to reclaim task %v...", task.TaskID)
				err := tsm.reclaim()
				if err != nil {
					log.Printf("ERROR: Encountered exception when reclaiming task %v - giving up retrying: %v", task.TaskID, err)
					return
				}
				log.Printf("Successfully reclaimed task %v", task.TaskID)
			}
		}
	}()
	return tsm
}

// stopReclaims() must be called when tsm.Lock() is held by caller
func (tsm *TaskStatusManager) stopReclaims() {
	if !tsm.finishedReclaiming {
		close(tsm.stopReclaiming)
		<-tsm.reclaimingDone
		tsm.finishedReclaiming = true
	}
}

File: workers/generic-worker-rs/src/taskstatus.rs
//! Task status management and reclaim loop.
//!
//! Manages the lifecycle of a task's status, including periodic reclaims
//! to extend the task's lease, and status change notifications.

use anyhow::Result;
use chrono::{DateTime, Utc};
use std::sync::Arc;
use tokio::sync::{watch, Mutex, RwLock};
use tokio::time::{self, Duration};

use crate::errors::{CommandExecutionError, TaskStatus, TaskUpdateReason};
use crate::tc::{HttpQueue, Queue};

/// Callback type for task status change listeners.
pub type StatusChangeCallback = Box<dyn Fn(TaskStatus) + Send + Sync>;

/// A registered status change listener.
pub struct TaskStatusChangeListener {
    pub name: String,
    pub callback: StatusChangeCallback,
}

/// Manages task status, reclaim loop, and status change notifications.
pub struct TaskStatusManager {
    task_id: String,
    run_id: u32,
    status: Arc<RwLock<TaskStatus>>,
    taken_until: Arc<RwLock<DateTime<Utc>>>,
    abort_exception: Arc<Mutex<Option<CommandExecutionError>>>,
    listeners: Arc<Mutex<Vec<Arc<TaskStatusChangeListener>>>>,
    stop_reclaiming_tx: Option<watch::Sender<bool>>,
    reclaim_handle: Option<tokio::task::JoinHandle<()>>,
}

impl TaskStatusManager {
    /// Create a new TaskStatusManager and start the reclaim loop.
    pub fn new(
        task_id: String,
        run_id: u32,
        initial_taken_until: DateTime<Utc>,
        queue: Arc<HttpQueue>,
    ) -> Self {
        let status = Arc::new(RwLock::new(TaskStatus::Claimed));
        let taken_until = Arc::new(RwLock::new(initial_taken_until));
        let abort_exception = Arc::new(Mutex::new(None));
        let listeners = Arc::new(Mutex::new(Vec::new()));
        let (stop_tx, stop_rx) = watch::channel(false);

        // Start reclaim loop
        let reclaim_handle = {
            let task_id = task_id.clone();
            let status = status.clone();
            let taken_until = taken_until.clone();
            let listeners = listeners.clone();

            tokio::spawn(async move {
                Self::reclaim_loop(
                    task_id, run_id, queue, status, taken_until, listeners, stop_rx,
                )
                .await;
            })
        };

        Self {
            task_id,
            run_id,
            status,
            taken_until,
            abort_exception,
            listeners,
            stop_reclaiming_tx: Some(stop_tx),
            reclaim_handle: Some(reclaim_handle),
        }
    }

    async fn reclaim_loop(
        task_id: String,
        run_id: u32,
        queue: Arc<HttpQueue>,
        status: Arc<RwLock<TaskStatus>>,
        taken_until: Arc<RwLock<DateTime<Utc>>>,
        listeners: Arc<Mutex<Vec<Arc<TaskStatusChangeListener>>>>,
        mut stop_rx: watch::Receiver<bool>,
    ) {
        loop {
            // Calculate next reclaim time (reclaim when 2/3 of time has passed)
            let until = { *taken_until.read().await };
            let now = Utc::now();
            let remaining = until.signed_duration_since(now);
            let sleep_duration = remaining.num_milliseconds().max(0) as u64 * 2 / 3;
            let sleep_duration = Duration::from_millis(sleep_duration.max(1000));

            tokio::select! {
                _ = time::sleep(sleep_duration) => {}
                _ = stop_rx.changed() => {
                    if *stop_rx.borrow() {
                        return;
                    }
                }
            }

            // Check if we should stop
            if *stop_rx.borrow() {
                return;
            }

            // Attempt reclaim
            match queue.reclaim_task(&task_id, run_id).await {
                Ok(response) => {
                    *taken_until.write().await = response.taken_until;
                    *status.write().await = TaskStatus::Reclaimed;

                    // Notify listeners
                    let listeners = listeners.lock().await;
                    for listener in listeners.iter() {
                        (listener.callback)(TaskStatus::Reclaimed);
                    }
                    tracing::info!(
                        "Successfully reclaimed task {}/{}, taken until {}",
                        task_id,
                        run_id,
                        response.taken_until
                    );
                }
                Err(e) => {
                    tracing::warn!("Failed to reclaim task {}/{}: {}", task_id, run_id, e);
                }
            }
        }
    }

    /// Get the current task status.
    pub async fn last_known_status(&self) -> TaskStatus {
        *self.status.read().await
    }

    /// Get the current taken_until time.
    pub async fn taken_until(&self) -> DateTime<Utc> {
        *self.taken_until.read().await
    }

    /// Get the abort exception, if any.
    pub async fn abort_exception(&self) -> Option<CommandExecutionError> {
        self.abort_exception.lock().await.take()
    }

    /// Register a status change listener.
    pub async fn register_listener(&self, listener: Arc<TaskStatusChangeListener>) {
        self.listeners.lock().await.push(listener);
    }

    /// Deregister a status change listener by name.
    pub async fn deregister_listener(&self, name: &str) {
        let mut listeners = self.listeners.lock().await;
        listeners.retain(|l| l.name != name);
    }

    /// Abort the task with the given error.
    pub async fn abort(&self, err: CommandExecutionError) -> Result<()> {
        let current_status = *self.status.read().await;
        match current_status {
            TaskStatus::Claimed | TaskStatus::Reclaimed => {
                *self.status.write().await = TaskStatus::Aborted;
                *self.abort_exception.lock().await = Some(CommandExecutionError::new(
                    err.task_status,
                    anyhow::anyhow!("{}", err.cause),
                    err.reason,
                ));

                let listeners = self.listeners.lock().await;
                for listener in listeners.iter() {
                    (listener.callback)(TaskStatus::Aborted);
                }
                Ok(())
            }
            _ => {
                tracing::warn!("Cannot abort task in status {}, ignoring", current_status);
                Ok(())
            }
        }
    }

    /// Cancel the task.
    pub async fn cancel(&self) -> Result<()> {
        *self.status.write().await = TaskStatus::Cancelled;
        let listeners = self.listeners.lock().await;
        for listener in listeners.iter() {
            (listener.callback)(TaskStatus::Cancelled);
        }
        Ok(())
    }

    /// Stop the reclaim loop.
    pub fn stop_reclaims(&mut self) {
        if let Some(tx) = self.stop_reclaiming_tx.take() {
            let _ = tx.send(true);
        }
    }

    /// Report the task as completed.
    pub async fn report_completed(&self, queue: &HttpQueue) -> Result<()> {
        self.stop_reclaim_if_needed();
        queue.report_completed(&self.task_id, self.run_id).await?;
        *self.status.write().await = TaskStatus::Succeeded;
        Ok(())
    }

    /// Report the task as failed.
    pub async fn report_failed(&self, queue: &HttpQueue) -> Result<()> {
        self.stop_reclaim_if_needed();
        queue.report_failed(&self.task_id, self.run_id).await?;
        *self.status.write().await = TaskStatus::Failed;
        Ok(())
    }

    /// Report the task as exception.
    pub async fn report_exception(
        &self,
        queue: &HttpQueue,
        reason: TaskUpdateReason,
    ) -> Result<()> {
        self.stop_reclaim_if_needed();
        queue
            .report_exception(&self.task_id, self.run_id, &reason.to_string())
            .await?;
        *self.status.write().await = TaskStatus::Errored;
        Ok(())
    }

    fn stop_reclaim_if_needed(&self) {
        if let Some(ref tx) = self.stop_reclaiming_tx {
            let _ = tx.send(true);
        }
    }
}

impl Drop for TaskStatusManager {
    fn drop(&mut self) {
        self.stop_reclaims();
        if let Some(handle) = self.reclaim_handle.take() {
            handle.abort();
        }
    }
}
